Assets/MRTK/Providers/LeapMotion/LeapMotionDeviceManager.cs

```diff
@@ -0,0 +1,242 @@
+        /// Adds a new LeapMotionArticulatedHand to the scene.
+        /// </summary>
+        /// <param name="handedness">The handedness (Handedness.Left or Handedness.Right) of the hand to be added</param>
+        private void OnHandDetected(Handedness handedness)
+        {
+            // Only create a new hand if the hand does not exist
+            if (!trackedHands.ContainsKey(handedness))
+            {
+                var pointers = RequestPointers(SupportedControllerType.ArticulatedHand, handedness);
+                var inputSource = CoreServices.InputSystem?.RequestNewGenericInputSource($"Leap {handedness} Controller", pointers, InputSourceType.Hand);
+                var leapHand = new LeapMotionArticulatedHand(TrackingState.Tracked, handedness, inputSource);
+
+                // Set the pointers for an articulated hand to the leap hand
+                foreach (var pointer in pointers)
+                {
+                    pointer.Controller = leapHand;
+                }
+
+                trackedHands.Add(handedness, leapHand);
+
+                CoreServices.InputSystem.RaiseSourceDetected(inputSource, leapHand);
+            }
+        }
+
+        /// <summary>
+        /// Removes the LeapMotionArticulated hand from the scene when the tracking is lost.
+        /// </summary>
+        /// <param name="handedness">The handedness (Handedness.Left or Handedness.Right) of the hand to be removed</param>
+        private void OnHandDetectionLost(Handedness handedness)
+        {
+            if (CoreServices.InputSystem != null)
+            {
+                CoreServices.InputSystem.RaiseSourceLost(trackedHands[handedness].InputSource);
+            }
+
+            // Remove the pointer gameobjects
+            // Should I be destroying the gameobjects or change the tracking state of the hand?
+            foreach (var pointer in trackedHands[handedness].InputSource.Pointers)
+            {
+                if (pointer != null && (MonoBehaviour)pointer != null && ((MonoBehaviour)pointer).gameObject != null)
+                {
+                    GameObject.Destroy(((MonoBehaviour)pointer).gameObject);
+                }
+            }
+
+            // Remove tracked hands
+            trackedHands.Remove(trackedHands[handedness].ControllerHandedness);
+        }
+
+        /// <summary>
+        /// Update the number of tracked leap hands.
+        /// </summary>
+        /// <param name="isLeftTracked">The tracking state of the left leap hand</param>
+        /// <param name="isRightTracked">The tracking state of the right leap hand</param>
+        private void UpdateLeapTrackedHands(bool isLeftTracked, bool isRightTracked)
+        {
+            // Left Hand Update
+            if (isLeftTracked && !trackedHands.ContainsKey(Handedness.Left))
+            {
+                OnHandDetected(Handedness.Left);
+            }
+            
+            if (!isLeftTracked && trackedHands.ContainsKey(Handedness.Left))
```

Comment:
Should these conditions be else-ifs? Or would it be valid for OnHandDetected and OnHandDetectionLost be valid to be called in the same run?

Reply:
The Update method updates the states of both the left and the right hand.  While it is not valid for `OnHandDetected` and `OnHandDetectionLost` to be called in the same run on the same hand.  It is valid for `OnHandDetected` and `OnHandDetectionLost` to be called for different hands at the same time.  

For example, if you have the left hand in frame and then you move the left hand out of frame as you move the right hand in frame.  There is a possibility that the tracking state of both the left and the right hand can change in the same frame, the chances of that happening are narrow, but still possible.

Reply:
Added else if for each hand